bigbluebutton-html5/imports/startup/server/customVirtualBackground.js

```diff
@@ -0,0 +1,160 @@
+const meteorRoot = realpathSync(`${process.cwd()}/../`);
+const env = Meteor.isDevelopment ? 'development' : 'production';
+const CUSTOM_BACKGROUND_DIR = process.env.BBB_HTML5_CUSTOM_VIRTUAL_BG
+  || '/var/www/bigbluebutton/client/images/virtual-backgrounds';
+const virtualBgDir = (env === 'development')
+  ? realpathSync(`${meteorRoot}'/../../../../public/resources/images/virtual-backgrounds/`)
+  : realpathSync(`${meteorRoot}/../programs/web.browser/app/resources/images/virtual-backgrounds/`);
+
+function cleanProgramJson() {
+  if (env === 'production') {
+    const programJsonPath = realpathSync(`${meteorRoot}/../programs/web.browser/program.json`);
+
+    if (existsSync(programJsonPath)) {
+      const mode = statSync(programJsonPath).mode;
+      chmodSync(programJsonPath, 0o700);
+      const programJsonData = readFileSync(programJsonPath);
+      const programJson = JSON.parse(programJsonData);
+
+      programJson.manifest = programJson.manifest.filter((config) => !config.custom);
+
+      writeFileSync(programJsonPath, JSON.stringify(programJson));
+      chmodSync(programJsonPath, mode);
+    }
+  }
+}
+
+// This will probably never faill
+try {
+  const currentVirtualBackgrounds = readdirSync(virtualBgDir, { withFileTypes: true });
+  currentVirtualBackgrounds.forEach((entry) => {
+    if (entry.isDirectory) return;
+    if (DEFAULT_VIRTUAL_BACKGROUNDS.indexOf(entry.name) === -1) {
+      rmSync(`${virtualBgDir}/${entry.name}`);
+      rmSync(`${virtualBgDir}/thumbnails/${entry.name}`, { force: true });
+    }
+  });
+  cleanProgramJson();
+} catch (error) {
+  // eslint-disable-next-line no-console
+  console.error('Error removing non-default virtual backgrounds.', error);
+}
+
+if (existsSync(CUSTOM_BACKGROUND_DIR)) {
+  const files = readdirSync(CUSTOM_BACKGROUND_DIR, 'utf-8');
+
+  if (files.length > 0) {
+    const configs = files.map((file) => {
+      try {
+        // Background
+        const bgPath = `${CUSTOM_BACKGROUND_DIR}/${file}`;
+        copyFileSync(bgPath, `${virtualBgDir}/${file}`);
+        spawnSync('jpegoptim', ['-q', `${virtualBgDir}/${file}`]);
+        const bgData = readFileSync(`${virtualBgDir}/${file}`, 'utf-8');
+        const bgStats = statSync(`${virtualBgDir}/${file}`);
+        const bgHash = createHash('sha1')
+          .update(bgData)
+          .digest('hex');
+        
+        Meteor.settings.public.virtualBackgrounds.fileNames.push(file);
+
+        // Thumbnail
+        const thumbPath = `${CUSTOM_BACKGROUND_DIR}/thumbnails/${file}`;
+        if (existsSync(thumbPath)) {
+          copyFileSync(thumbPath, `${virtualBgDir}/thumbnails/${file}`);
+        } else {
+          spawnSync('convert', [
+            `${CUSTOM_BACKGROUND_DIR}/${file}`,'-resize', '50x50^', '-gravity', 'center',
+            '-extent', '50x50', `${virtualBgDir}/thumbnails/${file}`
+          ]);
+        }
+        spawnSync('jpegoptim', ['-q', `${virtualBgDir}/thumbnails/${file}`]);
```

Comment:
If you run `jpegoptim` on files in `/var/www` they must be writable by user `meteor`, readable by `www-data`. Shouldn't optimization be left to the operator? This way operators could specify the acceptable quality loss.

Reply:
This is good point. I agree with you. My intention was to make the customization steps as easy as possible.